fix(chat): prevent GLM-4.5V infinite repetition with n > 1 completions

ayushsatyam146 opened the pull request:
Remove enable_thinking=False parameter to prevent dummy <think></think> tokens that cause infinite loops in GLM-4.5V when generating multiple completions. Unified fix logic across harmony and standard paths.

FIX #23438
FIX #23251

DarkLight1337 commented:
Hmm, does this mean that the issue is with the chat template rather than vLLM? cc @zRzRzRzRzRzRzR

david6666666 commented:
> @david6666666 if possible, can you share the `vllm serve ....` command for the model and also if possible can you once verify of it happens for n > 1 or n == 1 as well? I work on a mac mini and I am not able to reproduce this issue due to hardware constraints.

I use zai-org/GLM-4.1V-9B-Thinking instead
```
lm_eval --model vllm-vlm --model_args pretrained=zai-org/GLM-4.1V-9B-Thinking,tensor_parallel_size=4,max_model_len=32768,gpu_memory_utilization=0.9 --tasks chartqa --batch_size auto --apply_chat_template --limit 10 --log_samples --output_path ./results --verbosity DEBUG
```

ayushsatyam146 committed:
fix(chat): prevent GLM-4.5V infinite repetition with n > 1 completions

Remove enable_thinking=False parameter to prevent dummy <think></think>
tokens that cause infinite loops in GLM-4.5V when generating multiple
completions. Unified fix logic across harmony and standard paths.

Signed-off-by: Ayush Satyam <ayushsatyam146@gmail.com>

david6666666 commented:
> @david6666666 I pushed some new changes if possible can you check it out once more plase. Thanks a lot!

still issue
```
{
	"doc_id": 1,
	"doc": {
		"image": "c5f35dcca24e8c6d691425be32463d7649de5850958c079a52394fc82368db44",
		"query": "What is the difference in value between Lamb and Corn?",
		"label": ["0.57"],
		"human_or_machine": 0
	},
	"target": "0.57",
	"arguments": {
		"gen_args_0": {
			"arg_0": "[gMASK]<sop><|user|>\n<|begin_of_image|><|image|><|end_of_image|>What is the difference in value between Lamb and Corn?\nAnalyze the image and question carefully, using step-by-step reasoning.\nFirst, describe any image provided in detail. Then, present your reasoning. And finally your final answer in this format:\nFinal Answer: <answer>\nwhere <answer> follows the following instructions:\n- <answer> should should be a single phrase or number.\n- <answer> should not paraphrase or reformat the text in the image.\n- If <answer> is a ratio, it should be a decimal value like 0.25 instead of 1:4.\n- If the question is a Yes/No question, <answer> should be Yes/No.\n- If <answer> is a number, it should not contain any units.\n- If <answer> is a percentage, it should include a % sign.\n- If <answer> is an entity, it should include the full label from the graph.\nIMPORTANT: Remember, to end your answer with Final Answer: <answer>.\n<|assistant|>\n",
			"arg_1": {
				"until": [],
				"temperature": 0.0,
				"do_sample": false,
				"max_gen_toks": 512
			},
			"arg_2": {
				"visual": ["c5f35dcca24e8c6d691425be32463d7649de5850958c079a52394fc82368db44"]
			}
		}
	},
	"resps": [
		["<think>Got it, let's analyze the image and question carefully. First, I need to find the difference between Lamb and Corn. The image shows a bar chart showing the long-term price index for food commodities from 1850-2015, 1934. The question is asking for the difference between Lamb and Corn. First, I need to find the difference between Lamb and Corn. First, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. The Lamb value is 103.7 and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image,"]
	],
	"filtered_resps": ["<think>Got it, let's analyze the image and question carefully. First, I need to find the difference between Lamb and Corn. The image shows a bar chart showing the long-term price index for food commodities from 1850-2015, 1934. The question is asking for the difference between Lamb and Corn. First, I need to find the difference between Lamb and Corn. First, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. The Lamb value is 103.7 and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image, I need to get the values for Lamb and Corn. From the image,"],
	"filter": "none",
	"metrics": ["exact_match", "relaxed_accuracy", "anywhere_accuracy"],
	"doc_hash": "5e3fdffefcd12c2d329a55f9350094b1b55d765619bd3a0e72ccaa249813699d",
	"prompt_hash": "b2f7763fc6804de1af845cea9fe28020eaa59bdf6c29ee01dc9c1adc88ffca6f",
	"target_hash": "37f253d59e6203142bd9b0b9a7ac8044b7217baca69cfc9846e3748a3f804cc0",
	"exact_match": 0.0,
	"relaxed_accuracy": 0.0,
	"anywhere_accuracy": 0.0
}
```

ayushsatyam146 commented:
I think this can be an issue with model chat template as well but I have no further explanation or a possible solution to this. @david6666666 can you suggest something here if there is something you can notice?

ayushsatyam146 commented:
@DarkLight1337 if possible can you also help with any ideas please?

DarkLight1337 commented:
I doubt it is a problem with chat template, rather it has something do with sampler, since this is specific to `n > 1` and the chat template doesn't depend on that.
